Reworked tap-to-paginate

gmahomarf opened the pull request:
# Changed
- Changed: The Epub reader now has much better handling of highlighting when text is underneath the tap to paginate area. You can now highlight under it naturally. (Fixes #1931)

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1545,7 +1548,39 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+  maybeChangeCursor(event: MouseEvent) {
+    if (this.clickToPaginate) {
+      if (event.clientX <= window.innerWidth * 0.2 || event.clientX >= window.innerWidth * 0.8) {
+        this.pointer = true;
```
You need to toggle change detection here. It would be more ideal for this to be an observable that maps to a book from mouse move. This way we can denounce.

gmahomarf replied:
Got it. I'll take a look

gmahomarf replied:
I tried debouncing, but there was noticeable delays in changing the cursor. I decided on a different approach, still with Observables. See https://github.com/Kareadita/Kavita/pull/2061/files#diff-b77cd1f8004a52c328b463987f49e3ef7910992e272358a539b654f891960d33R480-R487

majora2007 reviewed (commented):
Just a quick look, the idea seems fine. I think we can streamline some of the code. I'll test it once I wrap some other work up and let you know.

majora2007 commented:
I'm planning to look into this later this week.

majora2007 commented:
I'm sure it's clear, but I actually moved this into next release due to some circumstances. It's at the top of the list for next release, so we can get it into an early nightly.

majora2007 commented:
I'm finally getting time to look into this. At first glance, it's awesome. You can easily scroll under the pagination area. However, I did notice some issues:
1. When clicking the navigation area quickly, sections of a paragraph could highlight (or single words)
2. The bottom action bar can flash on the screen

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -475,8 +477,14 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+        map(this.mapMousePositionToPointerCursor.bind(this)),
+        distinctUntilChanged())
+      .subscribe((pointer) => {
+        this.changeCursor(pointer);
```
This can be rewritten:
    fromEvent<MouseEvent>(this.bookContainerElemRef.nativeElement, 'mousemove')
      .pipe(
        takeUntilDestroyed(this.destroyRef),
        map(() => this.mapMousePositionToPointerCursor)),
        distinctUntilChanged(),
        tap((pointer) => this.changeCursor(pointer))
      .subscribe()

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1545,7 +1554,44 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+    this.cdRef.markForCheck();
+  }
 
+  handleReaderClick(event: MouseEvent) {
```
At first glance, you can rewrite this as:

```
handleReaderClick(event: MouseEvent) {
    if (!this.clickToPaginate) { return; }

    const probablyHighlighting = window.getSelection()?.toString() != '';

    console.log('probablyHighlighting: ', probablyHighlighting);
    if (event.clientX <= window.innerWidth * 0.2) {
      if (!probablyHighlighting) {
        this.movePage(this.readingDirection === ReadingDirection.LeftToRight ? PAGING_DIRECTION.BACKWARDS : PAGING_DIRECTION.FORWARD);
        return;
      }
    } else if (event.clientX >= window.innerWidth * 0.8) {
      if (!probablyHighlighting) {
        this.movePage(this.readingDirection === ReadingDirection.LeftToRight ? PAGING_DIRECTION.FORWARD : PAGING_DIRECTION.BACKWARDS);
        return;
      }
    }
    
    this.toggleMenu(event);
  }
  ```
  
  This employs Inversion of Control to avoid nesting on this.clickToPaginate. We can avoid likely buggy math by just getting the selection from the browser, that will tell us if we are actually selecting something. We can use shortcircuiting to avoid if/else nesting to keep the logic clean.

majora2007 replied:
Rewrote it to be even more clean:

```
handleReaderClick(event: MouseEvent) {
    if (!this.clickToPaginate) { return; }

    const isHighlighting = window.getSelection()?.toString() != '';
    if (isHighlighting) {
      this.toggleMenu(event);
      return;
    }

    if (event.clientX <= window.innerWidth * 0.2) {
      this.movePage(this.readingDirection === ReadingDirection.LeftToRight ? PAGING_DIRECTION.BACKWARDS : PAGING_DIRECTION.FORWARD);
    } else if (event.clientX >= window.innerWidth * 0.8) {
      this.movePage(this.readingDirection === ReadingDirection.LeftToRight ? PAGING_DIRECTION.FORWARD : PAGING_DIRECTION.BACKWARDS);
    }
  }
  ```

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -275,6 +275,8 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
 
   writingStyle: WritingStyle = WritingStyle.Horizontal;
 
+  pointer: boolean = false;
```
Add a JSDoc on top to indicate this controls if the cursor is a pointer or not based on if over a pagination area or not.

gmahomarf commented:
> I'm finally getting time to look into this. At first glance, it's awesome. You can easily scroll under the pagination area. However, I did notice some issues:
> 
> 1. When clicking the navigation area quickly, sections of a paragraph could highlight (or single words) 
You mean multiple clicks in a row? That's just the normal highlight behavior, no? Double-click selects a word, triple-click selects a line. Do you want to get rid of it? 

> 2. The bottom action bar can flash on the screen
Is this related to the previous one? As in, when selecting text, the action bar shows?

majora2007 reviewed (changes requested):
Please first take care of the merge and my comments on how the code is written. Then let's tackle the issues. 

I noticed also if you spam click, during that highlight, the next page wont load. It kinda gets stuck.

majora2007 commented:
Here is an example of the highlight issue. I'm always clicking on the right hand side of the screen.
![click issue](https://github.com/Kareadita/Kavita/assets/735851/57df6999-07e4-47a8-8a3f-13ac1f062da0)

majora2007 commented:
Just checking if you're interested in working to polish this off or not?

gmahomarf commented:
> Just checking if you're interested in working to polish this off or not?

Hey, sorry about that. It's been a busy few weeks for me and I haven't been able to make time. I _am_ interested in polishing it off, but I'm also not opposed to someone taking over, if there's a deadline or some other time concern.

majora2007 commented:
No deadline, just wanted to make sure, otherwise I would take it over, but it wouldn't be anytime soon as I'm booked on the next few releases. 

I'm going to just move this to draft so you can reopen when you're ready.

gmahomarf commented:
For unrelated reasons, I couldn't sleep last night. Since this was on my mind, here I am. I've addressed your formatting changes, and I believe I've tackled the text selection issues.

majora2007 commented:
I just started reviewing this and it's looking promising. I'll get this tested on my phone and tablet as well.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.html:
```diff
@@ -101,20 +101,21 @@ <h5 header>
       </ng-container>
-      <div #bookContainer class="book-container {{WritingStyleClass}}" [ngClass]="{'immersive' : immersiveMode}">
+        <div #bookContainer class="book-container {{WritingStyleClass}}"
+          [ngClass]="{'immersive' : immersiveMode, 'pointer' : pointer}"
```
This needs to be  `cursorIsPointer`

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -292,6 +292,19 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
   writingStyle: WritingStyle = WritingStyle.Horizontal;
 
+  /**
+   * Controls whether the cursor is a pointer when tap to paginate is on and the curso is over
```
Just a small typo too on cursor

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1568,7 +1588,38 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+      return;
+    }
+
+    if (event.clientX <= window.innerWidth * 0.2) {
```
What are these magic numbers? please put into a const just above so it's clear.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1581,17 +1632,27 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
-    this.mousePosition.y = $event.screenY;
+    this.mousePosition.x = $event.clientX;
+    this.mousePosition.y = $event.clientY;
+    if (!this.clickToPaginate || !this.isCursorOverPaginators($event)) {
```
Please leave a newline above the if to help separate the code.

majora2007 commented on UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:
```diff
@@ -1581,17 +1632,27 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
+
+    const now = Date.now();
+    // This prevents selecting text when clicking repeatedly to switch pages
+    if (now - this.lastPaginatorClickTime < 500) {
```
Put this magic number into a const. So debouncing wouldn't work on this?

gmahomarf replied:
Hmmm now that you mention it, this entire check is unnecessary. I'll remove it

gmahomarf replied:
Ok, this _may_ have been required for allowing starting a selection from a pagination area

majora2007 reviewed (changes requested):
Just a few changes required. I'll now give it a test on mobile devices to ensure it works appropriately, then we can merge.

majora2007 commented:
Looks good on mobile. Fix up the PR, I'll re-review on the PC then we can merge. This is a great improvement, thanks for taking it up.

gmahomarf commented:
> Looks good on mobile. Fix up the PR, I'll re-review on the PC then we can merge. This is a great improvement, thanks for taking it up.

I've addressed all comments, I believe. I'm more than happy to contribute to something I use frequently

majora2007 commented:
So I can highlight under the pagination area, but previously if there was text starting from under the pagination area, I was able to select them. Honestly, you might want to revert.

gmahomarf commented:
> So I can highlight under the pagination area, but previously if there was text starting from under the pagination area, I was able to select them. Honestly, you might want to revert.

Yep, I did just that. Everything should be back to the way it was, with the requested changes
